### ISSUE TITLE
Removes valid branch tracking information

### ISSUE THREAD

**Issue by jannisg** [Kikobeats/git-dirty#2]
I've just come across this package as I am setting up `bumped` for the first time on a new project but I'm running into an issue where

```
git branch --unset-upstream
```

on line 22 of the `dirty.sh` file is removing my tracking of `master origin/master`.

Given that `bumped` runs the following in sequence:

```
Detecting problems before publish:
      plugin: bumped-terminal
      command: 'git-dirty'
    Publishing tag to BitBucket:
      plugin: bumped-terminal
      command: 'git tag $newVersion && git push && git push --tags'
```

The `Publishing tag to BitBucket` always fails with

```
post bumped Starting Publishing tag to BitBucket
post bumped fatal: The current branch master has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin master

post bumped Process exited with code 128
```

As the `git-dirty` command seems to remove the tracking every time before the next step is running `git push`.

---

Aside from this repo being separate to `bumped` and this perhaps falling outside of this repo's responsibility, what is the need for the `git branch --unset-upstream` anyway?

It seems to me that in any use case it would always result in local tracking information being lost?
Maybe I'm missing something though…

Would love to hear your thoughts.

**Comment by Kikobeats** [comment]
hmmm not sure what is happening, but the final unset upstream is necessary because [at line 15](https://github.com/Kikobeats/git-dirty/blob/master/dirty.sh#L15) it set the branch temporaly as upstream.

Maybe we can use a differente mechanism for the same purpose?

**Comment by Kikobeats** [comment]
For example: https://stackoverflow.com/questions/1800783/compare-local-git-branch-with-remote-branch

**Comment by jannisg** [comment]
I think there are generally 2 options:

1. If we capture the original upstream setting in a variable (perhaps using `git branch -vv` output?), then rather than unsetting upstream we can set it back to the original value.
2. `git diff` provided it contains the info you need for the history check you're running at line 17.

I'd say, option 2 is cleaner but I'm not 100% sure what `git rev-list --count --left-only @'{u}'...HEAD` does/whether you can accomplish the same with `git diff`.

**Comment by Kikobeats** [comment]
I think we can use this command:

```
git diff master origin/master --exit-code
```

a more generic way is:

```
git diff master@{u}
```

where master is the currrent branch or provided by CLI

The point is, it is comparing with upstream and even you define a remote repository, if you don't setup it as upstream the command will fail.

**Comment by Kikobeats** [comment]
```
❯ g remotes
origin	git@github.com:windtoday/windtoday-app.git (fetch)
origin	git@github.com:windtoday/windtoday-app.git (push)

windtoday-app master
❯ g diff master@{u}
fatal: no upstream configured for branch 'master'
```

Actually I don't undertand why it doesn't undertand origin as default upstream...

**Comment by jannisg** [comment]
Hmm the above does work for me without errors.

Is it safe to assume someone is on `master`?

**Comment by jannisg** [comment]
We could parse the output of `git branch -vv` to get all info as well:

```
→ git branch -vv
* master c2950f2 [origin/master] Release 0.0.5
```

**Comment by Kikobeats** [comment]
nope, it was an example; it assumes the current branch by default.

About my problem, it was because I had setup push as `current`. Check here http://stackoverflow.com/a/948397.

We need a way to determinate the second diff parameter

**Comment by Kikobeats** [comment]
@jannisg sounds hacky buy possible 😄

**Comment by Kikobeats** [comment]
not possible, because If you haven't a upstream it doesn't output nothing:

```
❯ git branch -vv
* master 0eced3e Update flush
````

Thinking...

**Comment by jannisg** [comment]
Throw an error is no upstream is set at all? That seems more of a config issue that anything wouldn't it?

**Comment by Kikobeats** [comment]
At this moment of the conversation we are agree with replace the current behavior for:

```
git diff $CURRENT_BRANCH $REMOTE_BRANCH --exit-code
```

Exploring a way to have $REMOTE_BRANCH available

**Comment by Kikobeats** [comment]
@jannisg can you check that? http://stackoverflow.com/a/7733266

**Comment by jannisg** [comment]
```
→ git remote show origin
* remote origin
  Fetch URL: {myrepourl}
  Push  URL: {myrepourl}
  HEAD branch: master
  Remote branch:
    master tracked
  Local branch configured for 'git pull':
    master merges with remote master
  Local ref configured for 'git push':
    master pushes to master (up to date)
```

I wonder if this would be helpful, see the `git for-each-ref ` command http://stackoverflow.com/a/3633061/70339

**Comment by Kikobeats** [comment]
Hey @jannisg, I'm interest in resolve that, do you have any suggestion to go ahead? I'm a little out of the problem right now
